stdlib/public/core/Arrays.swift.gyb

```diff
@@ -1624,23 +1598,22 @@ extension ${Self} {
     return try body(&inoutBufferPointer)
   }
 
-  @discardableResult
   public func _copyContents(
-    initializing ptr: UnsafeMutablePointer<Element>
-  ) -> UnsafeMutablePointer<Element> {
-    if let s = self._baseAddressIfContiguous {
-      let count = self.count
-      ptr.initialize(from: s, count: count)
+    initializing buf: UnsafeMutableBufferPointer<Element>
+  ) -> Iterator? {
+    // FIXME: decide if this really is a fatalError
+    guard var p = buf.baseAddress 
+      else { fatalError("Attempt to copy contents into nil buffer pointer") }
+    if let s = _baseAddressIfContiguous {
+      p.initialize(from: s, count: self.count)
       _fixLifetime(self._owner)
-      return ptr + count
     } else {
-      var p = ptr
       for x in self {
         p.initialize(to: x)
         p += 1
       }
-      return p
     }
+    return nil
```

Comment:
This is the call into the appended collection, copying itself into the buffer of the appendee. The appended knows *it* didn't lie about it's count. The appendee is required to have allocated enough storage for it to append itself (though this should be documented as a precondition of initialize(from:), and checked here, will do that). This is different to the appendee collection, which can make no assumptions about the generic collection it received.

That said, based on benchmarks so far, this optimization might not be worth it.